### PR TITLE
Fix and refactor chart parsing behavior tests; implement Pro Keys parsing test

### PR DESCRIPTION
**Pull request by TheNathannator**
Main goal here was to improve maintainability; all of the manual sorted indexing into the various lists is difficult to reason about, very fragile, and scales quadratically in code complexity. Now, each type of event is added to a combined unsorted list, and then sorted afterwards, with no additional type queries being performed.

Sorting was already done for .mid by necessity, but it was not being taken advantage of to simplify things. Sorting was *not* done for .chart previously however, so that has been implemented to achieve the same maintenance simplicity.

Additionally, the amount of manual formatting done for writing out .chart files has been significantly reduced. The same code which handles the sorting also now does the writing, and outside code simply passes values to it.

Lastly, event ordering between the source/parsed charts are now ensured to be the same by sorting their contents before verifying them. This means notes/phrases on the same tick being in a different order will no longer result in a test failure. This sorting is only applied by the test by calling a new `Sort` function, so it will not affect the actual parsing time in-game.

### CODE REVIEW

**Comment by TheNathannator** [comment]
@sonicfind Anything you have comments on here?

@EliteAsian123 Anything that should be added to the Pro Keys tests?

**Comment by EliteAsian123** [comment]
Sorry for the late response, but, from the looks of it what you have there should be fine!

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs [hunk 1]
```diff
@@ -165,58 +181,33 @@ private static void GenerateInstrumentSection(MoonSong sourceSong, StringBuilder
...
             foreach (var phrase in phrasesToRemove)
             {
                 chart.Remove(phrase);
             }
 
-            builder.Append($"}}{NEWLINE}");
+            foreach (var text in chart.events.Concat(textPhrases))
+            {
+                section.AddEvent(text.tick, "E", text.text);
+            }
```
Comment (sonicfind) [inline YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs:207]: Those same three reasons applied for the old format of this code a while back too.

It's just that handling notes and phrases in this architecture has always been a little... undesirable.

File: YARG.Core/MoonscraperChartParser/MoonChart.cs [hunk 2]
```diff
@@ -83,6 +78,16 @@ public bool Remove(MoonText ev)
...
 
+        public void Sort()
+        {
+            notes.Sort((left, right) => left.InsertionCompareTo(right));
+            specialPhrases.Sort((left, right) => left.InsertionCompareTo(right));
+            events.Sort((left, right) => left.InsertionCompareTo(right));
+        }
+
+        // Only implemented to allow collection initializer support
+        IEnumerator IEnumerable.GetEnumerator() => throw new System.NotImplementedException();
```
Comment (sonicfind) [inline YARG.Core/MoonscraperChartParser/MoonChart.cs:89]: This screams more like language abuse than an actual good idea. `IEnumerable` is meant to signal that the type can be *enumerated* over. It's used to explicitly enable interactions with for-loops and LINQ functions. If you're expressively disallowing that functionality, then marking the type as enumerable is just lying.

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.cs [hunk 3]
```diff
@@ -21,418 +21,419 @@ internal class ParseBehaviorTests
...
-        );
+            NewNote(0, GuitarFret.Green),
+            NewNote(1, GuitarFret.Red),
+            NewNote(2, GuitarFret.Yellow),
+            NewNote(3, GuitarFret.Blue),
+            NewNote(4, GuitarFret.Orange),
+            NewNote(5, GuitarFret.Open),
+
+            NewSpecial(6, MoonPhrase.Type.Versus_Player1, length: 6),
+            NewNote(6, GuitarFret.Green, flags: Flags.Forced),
```
Comment (sonicfind) [inline YARG.Core.UnitTests/Parsing/ParseBehaviorTests.cs:47]: I don't really understand the benefit of removing the ParseBehavior type here. You're achieving the same thing that ParseBehavior provided but through a larger type, indirection from `Add` function calls that split the object types regardless, and utilization of dynamically-sized `Lists` instead of statically-sized `Arrays`.

Comment (sonicfind) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.cs:47]: And if this is why you needed to misuse `IEnumerable` for `MoonChart`, then I'd definitely recommend reversing course.

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs [hunk 4]
```diff
@@ -304,61 +336,55 @@ private static TrackChunk GenerateTrackChunk(MoonSong sourceSong, MoonInstrument
...
+
+            foreach (var phrase in chart.specialPhrases)
+            {
+                GenerateSpecialPhrase(events, phrase, gameMode);
+            }
+
+            foreach (var ev in chart.events)
+            {
+                events.Add((ev.tick, new MidiTextEvent(ev.text)));
+            }
```
Comment (sonicfind) [inline YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs:385]: I'm more fine with changing the insertion pattern for .midi, though. MidiEvent ordering will always hurt lol

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs [hunk 5]
```diff
@@ -175,6 +186,54 @@ private static Dictionary<int, int> BuildVocalsNoteLookup()
...
+            { MoonPhrase.Type.ProKeys_RangeShift5,   new[] { PRO_KEYS_SHIFT_5 } },
+            { MoonPhrase.Type.ProKeys_Glissando,     new[] { PRO_KEYS_GLISSANDO } },
+        };
+        #endregion
+
+        #region Lookups
+        private static readonly HashSet<MoonInstrument> SplitDifficultyTrackLookup = new()
+        {
+            MoonInstrument.ProKeys
+        };
```
Comment (sonicfind) [inline YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs:231]: I'd personally only use a hashset if we had 3+ instruments to handle with separated difficulty chunks. 2 or less just makes it a little unclear.

**Review by sonicfind — CHANGES_REQUESTED** [review]

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs [hunk 6]
```diff
@@ -17,13 +17,65 @@ namespace YARG.Core.UnitTests.Parsing
...
+                    return string.Compare(left.data, right.data, StringComparison.Ordinal);
+                });
+
+                foreach (var (tick, typeCode, data) in _events)
+                {
+                    _builder.Append($"  {tick} = {typeCode} {data}{NEWLINE}");
+                }
+
+                _builder.Append($"}}{NEWLINE}");
+            }
```
[the remark this answers is not in the document]
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs:69]: The intent of using the disposable pattern here was to avoid the need to manually finish the process, but in practice it's definitely more obtuse than doing it explicitly. In retrospect I'm not quite sure why I tend towards this kind of design lol

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs [hunk 7]
```diff
@@ -17,13 +17,65 @@ namespace YARG.Core.UnitTests.Parsing
...
+        }
+
```
[the remark this answers is not in the document]
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs:71]: Thinking as to how the .mid side of things is handled, yeah, there's no real need to group the list and string builder together like this.

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs [hunk 8]
```diff
@@ -499,25 +528,20 @@ private static void GenerateSpecialPhrase(MidiEventList events, MoonPhrase phras
...
-            events.Sort((ev1, ev2) => {
-                if (ev1.absoluteTick > ev2.absoluteTick)
-                    return 1;
-                else if (ev1.absoluteTick < ev2.absoluteTick)
-                    return -1;
+            events.Sort((left, right) =>
+            {
+                int compare = left.absoluteTick.CompareTo(right.absoluteTick);
+                if (compare != 0)
+                    return compare;
```
[the remark this answers is not in the document]
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs:535]: I disagree, at first glance I don't know what `left - right` means here without thinking about it for a good moment. Using `CompareTo` self-describes exactly what is intended, and ensures correct semantics by using a built-in function designed exactly for the task.

To this point, the [actual implementation of the function](https://github.com/dotnet/runtime/blob/88f9aba91e11d1695ebe9ab572736ee62ac7ad61/src/libraries/System.Private.CoreLib/src/System/Int64.cs#L81) doesn't use this shortcut due to overflow issues (which, while pedantic here, is absolutely something which could catch you off-guard when using it elsewhere):
image

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs [hunk 9]
```diff
@@ -78,80 +130,44 @@ public class ChartParseBehaviorTests
...
-                while (eventIndex < sourceSong.events.Count &&
-                    // Event comes before a section (equals does not count)
-                    (sectionIndex == sourceSong.sections.Count || sourceSong.events[eventIndex].tick < sourceSong.sections[sectionIndex].tick))
-                {
-                    var ev = sourceSong.events[eventIndex++];
-                    builder.Append($"  {ev.tick} = E \"{ev.text}\"");
-                }
+            foreach (var ev in sourceSong.events.Concat(sourceSong.sections))
+            {
+                section.AddEvent(ev.tick, "E", '"' + ev.text + '"');
```
[the remark this answers is not in the document]
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs:169]: It's significantly harder to read, understand, and maintain the code that you wrote, as the number of things to check scales quadratically as more and more event types are added. I care significantly more about comprehensibility here than trying to manually ensure everything is correct in a single pass.

If we want to test more specific edge cases, we should be writing more specific tests. These current tests are intended only as a general "make sure nothing blows up" test which covers all base facets of the parser. It's unreasonable to check everything within a single test.

Additionally, allocation efficiency and execution speed should not be priority goals with unit tests. I appreciate that you're thinking about that, but I don't care how long it takes to run them so long as it's not unreasonable lol, no more than a few seconds overall. (The performance of the code being tested is a different story, but that's what benchmarks are for.) We're also dealing with less than 100 events in all of our current data sets, so algorithmic complexity hardly matters in this scenario anyways.

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.cs (hunk 3 above)
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.cs:47]: What would be better here is for me to contain the language abuse to the unit tests instead of applying it to the core codebase. As I mentioned in one of my previous comments, my primary goal with these unit tests is comprehensibility, and splitting the notes and special phrases into separate arrays makes it significantly more difficult to see how the phrases relate to notes. You added comments as a stopgap, but having to remember to edit two things when we can reasonably only have to edit one just doesn't make sense to me.

File: YARG.Core/MoonscraperChartParser/MoonChart.cs [hunk 10]
```diff
@@ -83,6 +78,16 @@ public bool Remove(MoonText ev)
             return MoonObjectHelper.Remove(ev, events);
         }
 
+        public void Sort()
+        {
+            notes.Sort((left, right) => left.InsertionCompareTo(right));
+            specialPhrases.Sort((left, right) => left.InsertionCompareTo(right));
+            events.Sort((left, right) => left.InsertionCompareTo(right));
+        }
```
[the remark this answers is not in the document]
Comment (TheNathannator) [reply YARG.Core/MoonscraperChartParser/MoonChart.cs:86]: This sort is not intended to correct tick ordering issues, we can assert for ascending tick ordering before doing the sort to ensure that doesn't go unnoticed. The sort also doesn't need to exist in `MoonSong` and `MoonChart` directly, so I'll move it to the parse tests.

I've found having to dance around making the data be written in a specific order to match somewhat annoying, and I'd rather not deal with it if I don't have to. Below is not a case we need to care about, and I can't fix the ordering within the unit tests themselves just by modifying the list of events:
image

For the future, behavior which depends on specific event ordering in the file should be checked with dedicated tests, preferably using pre-written format-specific data rather than trying to force a generic system to replicate format-specific quirks.

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs [hunk 11]
```diff
@@ -304,61 +336,55 @@ private static TrackChunk GenerateTrackChunk(MoonSong sourceSong, MoonInstrument
...
                     continue;
 
                 var chart = sourceSong.GetChart(instrument, difficulty);
-
-                // Indexing the separate lists is the only way to
-                // 1: Not allocate more space for a combined list, and
-                // 2: Not rely on polymorphic queries
-                int noteIndex = 0;
-                int phraseIndex = difficulty == Difficulty.Expert? 0 : chart.specialPhrases.Count;
-                int eventIndex = difficulty == Difficulty.Expert ? 0 : chart.events.Count;
```
[the remark this answers is not in the document]
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs:324]: Good catch, didn't really notice that in the web of code that existed previously lol, and didn't think about it since I was trying to split out logic wherever possible.

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs [hunk 12]
```diff
@@ -499,25 +528,20 @@ private static void GenerateSpecialPhrase(MidiEventList events, MoonPhrase phras
...
-                    (NoteOnEvent on1, NoteOnEvent on2) =>
-                        on1.NoteNumber > on2.NoteNumber ? 1
-                        : on1.NoteNumber < on2.NoteNumber ? -1
-                        : 0,
-                    (NoteOffEvent off1, NoteOffEvent off2) =>
-                        off1.NoteNumber > off2.NoteNumber ? 1
-                        : off1.NoteNumber < off2.NoteNumber ? -1
-                        : 0,
+                    (NoteOnEvent on1, NoteOnEvent on2) => on1.NoteNumber.CompareTo(on2.NoteNumber),
+                    (NoteOffEvent off1, NoteOffEvent off2) => off1.NoteNumber.CompareTo(off2.NoteNumber),
```
[the remark this answers is not in the document]
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs:544]: I don't really find the logic to be clearer with that conditional chain. There is no longer an up-front indication for the relation between the two events, and you have to decipher the entire line and read into the conditional first before you can know that the false case for the ternary is doing double-duty as the default state when comparing against any other event type.

The pattern matching expresses both of these different states clearly, and I hardly consider it abuse (`switch` statements even allow you to elide the inner pair of parentheses when constructing a tuple as the condition; it's only necessary on switch expressions).

File: YARG.Core/MoonscraperChartParser/MoonChart.cs (hunk 2 above)
Comment (TheNathannator) [reply YARG.Core/MoonscraperChartParser/MoonChart.cs]: I moved the abuse to the unit tests and more clearly signposted its purpose. I otherwise have no plans on removing it, as keeping the phrases grouped together with the notes in the list is much more understandable than having them separate.

We do this same language abuse in main YARG for the settings menu (see `MetadataTab`), only there it *does* have an actual implementation due to the type in question using just a single list for storing the data being added. It makes things more concise than it otherwise would be with object initializer syntax (primarily reduces a level of nesting), and it enables that backing list to be private instead of public.

File: YARG.Core/MoonscraperChartParser/MoonChart.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (sonicfind) [reply YARG.Core/MoonscraperChartParser/MoonChart.cs]: > I've found having to dance around making the data be written in a specific order to match somewhat annoying

I agree. The fact that the tests require data ordering to be the exact same for both chart object causes a lot of issues; but that ends up being a problem with the *unit test*, **not** the data. If you’re having to manipulate the data in a way that would not occur in actual production just for a “pass”, then the test itself is at fault.

In general, the problem is simply that the tests don’t take into account that in-game, we allow events that occur on the same-tick to be in whatever order they wish. In the end, that ordering *should not matter*, as an event applied on a tick is still an applied on a tick. Therefore, we should instead alter the way the tests validate the data so that they pass regardless of their ordering.

Anything besides that means that the test itself is completely invalid.

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.cs (hunk 3 above)
Comment (sonicfind) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.cs]: The problem that arises is that this explicitly assumes that for the total lifetime of the project, all the note and phrase data will be in this exact hierarchy of objects. If we ever re-architect the note and phrase structures where they have essentially no relation to each other, then this wouldn’t work.

This was *my* attempt at some form of future proofing + generally cutting down having to explicitly tailor our actual data just for the unit tests.

Comment (sonicfind) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.cs]: You’d say that the ParseBehavior provides more extensibility vs. the combined list.

And while it may look “nice”, the fact of that matter is that the notes and phrases still get separated regardless.

File: YARG.Core/MoonscraperChartParser/MoonChart.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (TheNathannator) [reply YARG.Core/MoonscraperChartParser/MoonChart.cs]: The tests don't take same-tick order independence into account because it's much more difficult to check things in that manner. The only reason why order matters in this test is due to the usage of `CollectionAssert.AreEqual`, which does a strictly-linear comparison of every element in the collection.

If event ordering generally doesn't matter, I don't think it's necessary to require that events are compared in an order-independent manner for the general test. As I've mentioned already, if there are specific cases to test to ensure this behavior, we should write specific tests for that instead of trying to shove it all into one test.

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.cs (hunk 3 above)
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.cs]: > If we ever re-architect the note and phrase structures where they have essentially no relation to each other

Under what scenario would this happen? Phrases will *always* inherently relate to notes and other events on the track, by nature of the genre we're designing for. They're only separate because combining them all into one list doesn't make sense for our purposes. And if we re-architect the chart data to work differently to how it does now, we'd have to rework more than just these unit tests. We should not be future-proofing for drastic structure changes.

> the notes and phrases still get separated regardless.

That doesn't mean we have to explicitly show that and destroy the relation between events here. If multiple things happen on the same tick, that should be explicitly visible in this list. You don't edit notes and phrases in completely separate views in a chart editor, so why should we here?

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs (hunk 5 above)
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs:231]: By "aggressive" I mean it's not actually deliberately future-proofing for anything particular, it is purely a decision I made based on the semantics of how the data is used: a lookup table of unique values. Whether we use a list or a hash map here doesn't matter in the long run, but a hash map is more suited for this particular use case in theoretical terms.

The only other instrument I can think of which would also require this is Phase Shift's Real Keys tracks, if we ever decide to read that in.

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (sonicfind) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs:549]: Ah fair enough

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs [hunk 13]
```diff
@@ -17,13 +17,65 @@ namespace YARG.Core.UnitTests.Parsing
...
+
+            public void AddEvent(uint tick, string typeCode, uint value1)
+            {
+                _events.Add((tick, typeCode, $"{value1}"));
+            }
+
+            public void AddEvent(uint tick, string typeCode, uint value1, uint value2)
+            {
+                _events.Add((tick, typeCode, $"{value1} {value2}"));
+            }
```
[the remark this answers is not in the document]
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs]: The point of these functions is to avoid having to repeat the same string interpolations over and over, which could potentially be error-prone. It's much easier to think about methods that have strong types and parameters than it is to manually construct the same category of strings. Using format strings like Moonscraper does for its writing is still potentially error-prone due to the lack of strong typing.

Just because it's not absolutely necessary doesn't mean it has no benefit.

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs [hunk 14]
```diff
@@ -17,13 +17,65 @@ namespace YARG.Core.UnitTests.Parsing
 
     public class ChartParseBehaviorTests
     {
+        private class ChartSection : IDisposable
+        {
+            private StringBuilder _builder;
+            private List<(uint tick, string typeCode, string data)> _events = new();
```
[the remark this answers is not in the document]
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Chart.cs]: Got them combined.

File: YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (sonicfind) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs:550]: wait actually, read that comment again:
> subtraction will wrap to position for very large [**negative**] numbers

We never have any tick values in the negative range, so that problem doesn't exist for our use.
Even if it was `0 - int.MaxValue`, the operation would still be correct

[the remark this answers is not in the document]
Comment (TheNathannator) [reply YARG.Core.UnitTests/Parsing/ParseBehaviorTests.Midi.cs:550]: Even so, I still disagree with this being more clear. It is a little more compact, yes, but the explicit naming of `CompareTo` alone makes it immediately clear what's going on.

Additionally, this trick does not work for unsigned integers. You cannot distinguish between `(int)(0u - 0x7FFFFFFFu)` and `(int)(0xFFFFFFFFu - 0x80000000u)`, for example, and the result is incorrect for half of the unsigned integer range. Rather than trying to dance around these pitfalls, it's much easier to use built-in functionality which is tested and proven to do the job correctly.

Whether we would actually hit the problem cases here is a different story, but this oversteps my personal safety line regarding programming for a specific use case, and just seems like using a trick to micro-optimize code where it's not necessary to (no offense intended). I'd much rather over-protect against possible issues and never actually hit them, than take too many shortcuts and cause issues later down the line.
